contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,491 @@
+it makes great sense to ensure that the instances are distributed into maximum clusters possible.
+The initial proposal is to simply hash the replica number into the number of healthy clusters.
+A simple modulo calculation can be used to do the distribution.
+
+### Instance identity and discovery
+
+In the case of in-cluster statefulset, pods discover each other using the in-cluster dns names.
+A headless service with selectors enables creation of dns records against the pod names.
+This cannot work across clusters, as local pod IPs cannot be reached across the clusters.
+
+Similar to the primary alternative, the proposal to make it work across the clusters is to
+assign a service type &#39;LoadBalancer&#39; for each pod instance that is created in the clusters.
+The service name will be based on the ordinal pod name assigned to each pod.
+Additionally a set of label selectors will be added to each individual pod to ensure that the
+cluster routes the specific service traffic to the specific pod.
+The service type &#39;LoadBalancer&#39; will ensure that the data traffic can be routed across clusters.
+The network dns names for each pod will remain very similar to the network names used by the in-cluster statefulsets.
+
+As an example (using the same yamls and names as in the earlier par of this doc):
+
+For a federation with 3 clusters all in the same DNS zone _example.com_ and the statefulset
+name _store_ the dns names for instances might look like:
+```
+store-0.etcd.mynamespace.myfederation.svc.federation.example.com
+store-1.etcd.mynamespace.myfederation.svc.federation.example.com
+store-2.etcd.mynamespace.myfederation.svc.federation.example.com
+store-3.etcd.mynamespace.myfederation.svc.federation.example.com
+store-4.etcd.mynamespace.myfederation.svc.federation.example.com
+store-5.etcd.mynamespace.myfederation.svc.federation.example.com
+store-6.etcd.mynamespace.myfederation.svc.federation.example.com
+store-7.etcd.mynamespace.myfederation.svc.federation.example.com
+store-8.etcd.mynamespace.myfederation.svc.federation.example.com
+store-9.etcd.mynamespace.myfederation.svc.federation.example.com
+store-10.etcd.mynamespace.myfederation.svc.federation.example.com
+store-11.etcd.mynamespace.myfederation.svc.federation.example.com
+```
+This will be achieved by federated controller creating and monitoring individual services in
+specific clusters against each pod, with the same name as that pod&#39;s
+ordinal identity. It will further need to be ensured that the pods be assigned the same hostname
+as the pod name (ordinal identity) and service name to keep it consistent with the in-cluster
+statefulsets topology.
+
+The details about creation and handling of the LB services and updating the dns records will
+be similar to that elaborated in the previous design.
+
+This alternative has following drawbacks:
+ - As the control over the stateful pods is with the federation statefulset controller, the
```

Comment:
This limitation could be overcome if the federated statefulsets controller delegated responsibility for monitoring stateful pods to a federation-specific controller in each targeted member cluster.  I think this proposal needs to be fleshed out more before it is disqualified as not being viable.